### PR TITLE
[SPARK-43945][SQL][TESTS] Fix bug for `SQLQueryTestSuite` when run on local env

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to fix bug for SQLQueryTestSuite when run on local env.
When SQLQueryTestSuite runs `identifier-clause.sql` in various different environments, the statement `DESCRIBE SCHEMA IDENTIFIER('id' || 'ent')` will generate different `Owner` values. Currently, this value(`runner`) is OK when runs on GA, but it will fail on local. In order to be compatible with the results of various different environments, the `Owner` value is removed for comparison.

### Why are the changes needed?
Fix bugs and compatibility.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Manually test.
- Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
[info] - identifier-clause.sql *** FAILED *** (809 milliseconds)
[info]   identifier-clause.sql
[info]   Expected "...ce Name   ident
[info]   Owner  [runner]", but got "...ce Name  ident
[info]   Owner  [panbingkun]" Result did not match for query #61
[info]   DESCRIBE SCHEMA IDENTIFIER('id' || 'ent') (SQLQueryTestSuite.scala:842)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.funsuite.AnyFunSuite.newAssertionFailedException(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Assertions.assertResult(Assertions.scala:847)
[info]   at org.scalatest.Assertions.assertResult$(Assertions.scala:842)
[info]   at org.scalatest.funsuite.AnyFunSuite.assertResult(AnyFunSuite.scala:1564)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.$anonfun$readGoldenFileAndCompareResults$3(SQLQueryTestSuite.scala:842)
[info]   at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[info]   at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[info]   at scala.collection.AbstractIterable.foreach(Iterable.scala:926)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.readGoldenFileAndCompareResults(SQLQueryTestSuite.scala:833)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.$anonfun$runQueries$10(SQLQueryTestSuite.scala:586)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]   at org.scalatest.Assertions.withClue(Assertions.scala:1065)
[info]   at org.scalatest.Assertions.withClue$(Assertions.scala:1052)
[info]   at org.scalatest.funsuite.AnyFunSuite.withClue(AnyFunSuite.scala:1564)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.runQueries(SQLQueryTestSuite.scala:582)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.$anonfun$runSqlTestCase$35(SQLQueryTestSuite.scala:424)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.$anonfun$runSqlTestCase$35$adapted(SQLQueryTestSuite.scala:422)
[info]   at scala.collection.immutable.List.foreach(List.scala:333)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.runSqlTestCase(SQLQueryTestSuite.scala:422)
[info]   at org.apache.spark.sql.SQLQueryTestSuite.$anonfun$createScalaTestCase$6(SQLQueryTestSuite.scala:329)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:221)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:67)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
[info]   at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:67)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:333)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:67)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:67)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:413)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:750)

**Comment by HyukjinKwon** [comment]
What's the bug? Mind elabourating it in the PR description?

**Comment by LuciferYang** [comment]
@panbingkun Can we just enhance `SQLQueryTestHelper#replaceNotIncludedMsg`, replace `Owner.*` to `s"Owner $notIncludedMsg"`?

https://github.com/apache/spark/blob/6f593beaf2f527f805da4463c19a89a8d9eea619/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala#L42-L57

And  I think this one worth file a new Jira

**Comment by beliefer** [comment]
@panbingkun Thank you for the fix. It seems it's not the issue of `SQLQueryTestSuite`. Could we find out the root cause ?

**Comment by panbingkun** [comment]
> What's the bug? Mind elabourating it in the PR description?

Ok, Let me do it.

**Comment by panbingkun** [comment]
> @panbingkun Can we just enhance `SQLQueryTestHelper#replaceNotIncludedMsg`, replace `Owner.*` to `s"Owner $notIncludedMsg"`?
> 
> https://github.com/apache/spark/blob/6f593beaf2f527f805da4463c19a89a8d9eea619/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala#L42-L57
> 
> And I think this one worth file a new Jira

Ok, Very good proposal, let me do it.

**Comment by LuciferYang** [comment]
> @panbingkun Thank you for the fix. It seems it's not the issue of `SQLQueryTestSuite`. Could we find out the root cause ?

`owner` is namespace reserved properties when `spark.sql.legacy.notReserveProperties` is false,  it always use `SPARK_USER` or `UserGroupInformation.getCurrentUser().getShortUserName()` and it cannot be modified

**Comment by beliefer** [comment]
> SPARK_USER

Thank you for your investigation. If so, it seems we just set the `SPARK_USER` or `CurrentUser` in afterAll.

**Review by dongjoon-hyun — COMMENTED** [review]
It looks worthy for a JIRA. please file a JIRA for this stuff for trace-ability, @panbingkun .

**Comment by panbingkun** [comment]
> It looks worthy for a JIRA. please file a JIRA for this stuff for trace-ability, @panbingkun .

OK.

**Comment by beliefer** [comment]
@panbingkun I tested with the below command and it works fine.
`SPARK_GENERATE_GOLDEN_FILES=1 SPARK_USER="runner" build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z identifier-clause.sql"`

I think we should just update the document of `org.apache.spark.sql.SQLQueryTestSuite` and add SPARK_USER="runner" in usage section.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 1]
```diff
@@ -838,13 +837,17 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
...
+      assertResult(normalizeOwner(expected.output), s"Result did not match" +
         s" for query #$i\n${expected.sql}") {
         output.output
       }
     }
   }
 
+  private def normalizeOwner(plan: String): String = {
+    plan.replaceAll("""Owner\t\{runner}""", "Owner\t" + Utils.getCurrentUserName)
+  }
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:849]: I know this work is OK. But it introduce the extra complexity and readability.

**Comment by LuciferYang** [comment]
> @panbingkun I tested with the below command and it works fine. `SPARK_GENERATE_GOLDEN_FILES=1 SPARK_USER="runner" build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z identifier-clause.sql"`
> 
> I think we should just update the document of `org.apache.spark.sql.SQLQueryTestSuite` and add `SPARK_USER="runner"` in usage section.

Export `SPARK_USER="runner"` in `build/mvn` and `build/sbt` as default ? This is easier for developers

**Comment by panbingkun** [comment]
This is all done! @HyukjinKwon @dongjoon-hyun @LuciferYang @beliefer

**Comment by beliefer** [comment]
> 
> Export `SPARK_USER="runner"` in `build/mvn` and `build/sbt` as default ? This is easier for developers

Seems fine.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala [hunk 2]
```diff
@@ -49,6 +49,7 @@ trait SQLQueryTestHelper extends Logging {
       .replaceAll("Created By.*", s"Created By $notIncludedMsg")
       .replaceAll("Created Time.*", s"Created Time $notIncludedMsg")
       .replaceAll("Last Access.*", s"Last Access $notIncludedMsg")
+      .replaceAll("Owner.*", s"Owner $notIncludedMsg")
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:52]: This change seems good to me too. @panbingkun Thank you!

**Comment by LuciferYang** [comment]
> > @panbingkun I tested with the below command and it works fine. `SPARK_GENERATE_GOLDEN_FILES=1 SPARK_USER="runner" build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z identifier-clause.sql"`
> > I think we should just update the document of `org.apache.spark.sql.SQLQueryTestSuite` and add `SPARK_USER="runner"` in usage section.
> 
> Export `SPARK_USER="runner"` in `build/mvn` and `build/sbt` as default ? This is easier for developers

@panbingkun can you test this change? May need additional check for Maven's testing

**Review by beliefer — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
Let's waiting ci ~

**Review by HyukjinKwon — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @dongjoon-hyun @beliefer @HyukjinKwon @LuciferYang  for review.

**Pull request closed without merging**